src/bin/pg_autoctl/supervisor.c

```diff
@@ -56,22 +72,277 @@ static bool supervisor_restart_service(Supervisor *supervisor,
 									   Service *service,
 									   int status);
 
+static void supervisor_handle_failed_restart(Supervisor *supervisor,
+											 Service *service,
+											 int status);
+
 static bool supervisor_may_restart(Service *service);
 
 static bool supervisor_update_pidfile(Supervisor *supervisor);
 
+static int supervisor_handle_dynamic(Supervisor *supervisor);
+
+
+/*
+ * supervisor_it_* implement a helper to loop through the active services in a
+ * supervisor struct. They can be used either on their own, or via the helper
+ * macro foreach_supervised_service().
+ * Consult the macro for an example of use.
+ */
+static inline Service *
+supervisor_it_get(Supervisor_it *sit)
+{
+	if (sit->index < sit->count)
+	{
+		return &sit->services[sit->index];
+	}
+
+	return NULL;
+}
+
+
+static inline Service *
+supervisor_it_first(Supervisor_it *sit, Supervisor *supervisor)
+{
+	*sit = (Supervisor_it) {
+		.supervisor = supervisor,
+		.services = supervisor->services,
+		.count = supervisor->serviceCount,
+		.index = 0,
+	};
+
+	return supervisor_it_get(sit);
+}
+
+
+static inline Service *
+supervisor_it_next(Supervisor_it *sit)
+{
+	sit->index++;
+
+	/* if we are done with the static services, move to the dynamic */
+	if (sit->index >= sit->count &&
+		sit->services == sit->supervisor->services)
+	{
+		sit->services = sit->supervisor->dynamicServices.services;
+		sit->count = sit->supervisor->dynamicServices.serviceCount;
+		sit->index = 0;
+	}
+
+	return supervisor_it_get(sit);
+}
+
+
+#define foreach_supervised_service(it, supervisor, service) \
+	for ((service) = supervisor_it_first((it), (supervisor)); \
+		 (service) != NULL; \
+		 (service) = supervisor_it_next(it))
+
+
+/*
+ * supervisor_add_dynamic starts the service and if successfull, adds it to the
+ * dynamicService array.
```

Comment:
What about `supervisor_dynamic_service_enable` as a function name? with the same prefix, and more clarity that the service is going to be started rather than just added to our array.